### PR TITLE
[terra-functional-testing] Setup Initial WDIO 6 Test Runner

### LINKED ISSUE #405: [terra-functional-testing] Setup initial WDIO test runner
# Feature Request

Setup a test runner and executable bin/script command in the terra-functional-testing package.

## Description

Create and setup an executable bin/script command for invoking the WebdriverIO 6 test runner. The initial test runner should use the default mocha framework and chromedriver recommended by WebdriverIO. 

https://webdriver.io/docs/gettingstarted.html

The executable bin command may need a new naming convention. The existing test runner is invoked using `tt-wdio`.

### PR DESCRIPTION
**Pull request by StephenEsser**
### Summary

This PR sets up the initial WebDriverIO 6 test runner using the defaults and recommendations on the [WebDriverIO Getting Started page](https://webdriver.io/docs/gettingstarted.html).

This PR is intended to setup the anticipated structure of the project and will be built upon to integrate the selenium and terra services.

An example project consuming these changes can be demoed [here](https://github.com/StephenEsser/test-runner-example). The project runs the auto-generated wdio test example using the terra-functional-testing package test runner.

Resolves: https://github.com/cerner/terra-toolkit/issues/405

### Additional Details

This PR also includes the following:

- Added jest setup at the root level of the project.
- Added link-parent-bin dependency to expose jest commands to sub-packages.
- Reactivated jest testing on travis
- Removed babel references. Currently, this project does not need or use babel. Babel can be reintroduced in the future if/when necessary.

Miscellaneous: After installing jest I used the `jest --init` command to create a jest configuration. The default filename is generated as `jest.config.js`. This is the reason for the jest setup rename. 

@cerner/terra

[CONTRIBUTORS.md]: ../blob/master/CONTRIBUTORS.md
[License]: ../blob/master/LICENSE

### CODE REVIEW

File: packages/terra-functional-testing/lib/cli/index.js
Comment (emilyrohrbough) [inline packages/terra-functional-testing/lib/cli/index.js:1]: I'm curious, why use this file as an import-export pass through? Why not use this file for the CLI logic as well?

File: packages/terra-functional-testing/lib/test-runner/wdio.conf.js
Comment (emilyrohrbough) [inline packages/terra-functional-testing/lib/test-runner/wdio.conf.js]: Can we move to config dir?
```
packages/terra-functional-testing/config/wdio.config.js
```

File: packages/terra-functional-testing/lib/test-runner/test-runner.js
Comment (emilyrohrbough) [inline packages/terra-functional-testing/lib/test-runner/test-runner.js:3]: Should we wrap the wdio CLI or should we use the wdio launcher directly?

File: packages/terra-functional-testing/lib/cli/test-runner-cli.js
Comment (emilyrohrbough) [inline packages/terra-functional-testing/lib/cli/test-runner-cli.js]:
```suggestion
    .option('-c, --config <path>', 'A file path to the test runner configuration.');
```

File: packages/terra-functional-testing/lib/test-runner/test-runner.js
Comment (emilyrohrbough) [reply packages/terra-functional-testing/lib/test-runner/test-runner.js:3]: Are you doing this to expose the CLI commands without redefining them?

File: packages/terra-functional-testing/lib/test-runner/test-runner.js
Comment (emilyrohrbough) [inline packages/terra-functional-testing/lib/test-runner/test-runner.js]: can we namespace with terra-functional-testing or the script name we decide on?

File: packages/terra-functional-testing/lib/test-runner/test-runner.js
Comment (emilyrohrbough) [inline packages/terra-functional-testing/lib/test-runner/test-runner.js:32]: Is it possible to use promise syntax here?

File: packages/terra-functional-testing/lib/test-runner/wdio.conf.js
Comment (emilyrohrbough) [inline packages/terra-functional-testing/lib/test-runner/wdio.conf.js]: We will want to scale this back to 1 in CI

File: packages/terra-functional-testing/lib/test-runner/wdio.conf.js
Comment (emilyrohrbough) [inline packages/terra-functional-testing/lib/test-runner/wdio.conf.js]: We might need to investigate impact grid infrastructure..unfortunately not sure it could handle the load, though locally it might be okay! It would speed up testing

This config here is related to the issue I pinged you about in regards to v4 loading the wdio.config multiple times...each spawned process was reloading the wdio config 1+ times. I'm really hoping this was fixed in webdriver v5 or v6

File: packages/terra-functional-testing/lib/test-runner/wdio.conf.js
Comment (emilyrohrbough) [inline packages/terra-functional-testing/lib/test-runner/wdio.conf.js]: Did webdriver v5/v6 drop support for JSONWireFrame protocol commands? In v4 set `logLevel=silent` to swallow these deprecation logs for commands that did not have a replacement: 

Here is an example of the v4 command `browser.sessions(): https://github.com/webdriverio-boneyard/v4/blob/86a1cbd33b386d06a18c45fb0e8d676699bd14c7/lib/protocol/sessions.js#L25

looking through v4 API it does look like most commands do now have a replacement. If I remember correctly, one of the biggest blocked on using the last W3C protocol commands was chrome lacking the actions support, which I think they have added in now...

File: packages/terra-functional-testing/package.json
Comment (emilyrohrbough) [inline packages/terra-functional-testing/package.json:42]: Do we plan to expand on this as we expand on the capabilities of this package?

File: package.json
Comment (emilyrohrbough) [inline package.json:25]: did we want to drop this?

File: browserslist-config-terra.js
Comment (emilyrohrbough) [inline browserslist-config-terra.js:1]: did we want to drop this?

**Review by emilyrohrbough — COMMENTED** [review]
Overall changes look good to me!

I know I made a few comments on the configuration & I know you mentioned you populated with the default to get us started. These comments are just thoughts I had while reading through that we could looking later or drop depending on the upgrade needs/usage. Just didn't want to forget about them.

File: packages/terra-functional-testing/lib/cli/index.js
Comment (StephenEsser) [reply packages/terra-functional-testing/lib/cli/index.js:1]: I think responsibilities of the CLI may continue to grow. I like the consistency of the directories, index files, and specific files behaving in an anticipated way

File: packages/terra-functional-testing/lib/test-runner/wdio.conf.js
Comment (StephenEsser) [reply packages/terra-functional-testing/lib/test-runner/wdio.conf.js]: I'm open to moving these files around. Is there strong reasoning for moving this file or is it to make it consistent with other existing packages?

File: packages/terra-functional-testing/lib/test-runner/test-runner.js
Comment (StephenEsser) [reply packages/terra-functional-testing/lib/test-runner/test-runner.js:3]: The default export is just the launcher. Only the launcher is being used here:

https://github.com/webdriverio/webdriverio/blob/master/packages/wdio-cli/src/index.js#L76

Does that answer your question?

File: packages/terra-functional-testing/lib/test-runner/test-runner.js
Comment (StephenEsser) [reply packages/terra-functional-testing/lib/test-runner/test-runner.js:32]: Promises and async/awaits are interchangeable under most circumstances. I chose async/await here for the explicit and newer syntax.

File: packages/terra-functional-testing/lib/test-runner/wdio.conf.js
Comment (StephenEsser) [reply packages/terra-functional-testing/lib/test-runner/wdio.conf.js]: I'll start a list for changes like these. I want to document anything we choose to alter from the defaults.

File: packages/terra-functional-testing/lib/test-runner/wdio.conf.js
Comment (emilyrohrbough) [reply packages/terra-functional-testing/lib/test-runner/wdio.conf.js]: For me, a few reasons
1. consistency with other packages
2. lib directories make me think of compiled / ship api plus we separate our tests from lib so separating config from config makes sense to me
3. less to write when using int eh command line like: `node_modules/terra-fuctional-testing/config/wdio.config` 

I also like knowing where to find the configs instead of having to dig through nested dirs on git-hub to find a file ref. that's a personal preference 😄

File: package.json
Comment (StephenEsser) [reply package.json:25]: Yes, unfortunately this wasn't working with babel. I started an issue thread with one of the maintainers of browserslist here: https://github.com/browserslist/browserslist/issues/484

We'll likely need to duplicate the browserslist config (or use the env) if and when we ever need it at this level. Right now we're not using it for anything so I removed it

File: browserslist-config-terra.js
Comment (StephenEsser) [reply browserslist-config-terra.js:1]: Yes, unfortunately this wasn't working as intended. A little more explanation on this comment:

https://github.com/cerner/terra-toolkit/pull/406#discussion_r420970349

File: packages/terra-functional-testing/lib/test-runner/test-runner.js
Comment (emilyrohrbough) [reply packages/terra-functional-testing/lib/test-runner/test-runner.js:3]: yes that does, I had assumed the launcher was it's own package.

File: packages/terra-functional-testing/lib/test-runner/test-runner.js
Comment (yuderekyu) [inline packages/terra-functional-testing/lib/test-runner/test-runner.js:36]: as we add more cli flags in the future, will we pass options as the 2nd parameter?
```suggestion
      const testRunner = new Launcher(configPath, options);
```

Comment (StephenEsser) [reply packages/terra-functional-testing/lib/test-runner/test-runner.js:36]: Yes, as more options are added they will be passed to the launcher. The options will need to be extracted and organized so we only pass options relevant to the wdio launcher. We don't want to pass the entire options parameter to the launcher as it will likely include custom options from our cli interface.

File: packages/terra-functional-testing/lib/test-runner/test-runner.js
Comment (StephenEsser) [reply packages/terra-functional-testing/lib/test-runner/test-runner.js]: I'll hard code a namespace for now. Long term we might want to handle these messages differently. Our Logger or one of the few debugging libraries available might be a good fit

File: packages/terra-functional-testing/lib/cli/test-runner-cli.js
Comment (StephenEsser) [reply packages/terra-functional-testing/lib/cli/test-runner-cli.js]: Added the shorthand cli option:

https://github.com/cerner/terra-toolkit/pull/406/commits/323a62a599edd71f5c21abcd5ba004bf9b8370eb

File: packages/terra-functional-testing/lib/test-runner/test-runner.js
Comment (StephenEsser) [reply packages/terra-functional-testing/lib/test-runner/test-runner.js]: Namespaced the console message:
https://github.com/cerner/terra-toolkit/pull/406/commits/323a62a599edd71f5c21abcd5ba004bf9b8370eb

File: packages/terra-functional-testing/lib/test-runner/wdio.conf.js
Comment (StephenEsser) [reply packages/terra-functional-testing/lib/test-runner/wdio.conf.js]: I moved the wdio.conf.js into the config directory. I might change my mind on this decision when I start integrating the selenium configurations. I'm a little conflicted about the services and configurations that support those services being disconnected in the file structure, but we can start those discussions when we get there.

File: packages/terra-functional-testing/package.json
Comment (StephenEsser) [reply packages/terra-functional-testing/package.json:42]: Potentially, this will be evaluated as we begin to introduce the selenium services and docker. We may end up removing this entirely depending on how our testing strategy evolves (Locally, CI, docker).

File: packages/terra-functional-testing/lib/test-runner/wdio.conf.js
Comment (mjhenkes) [reply packages/terra-functional-testing/lib/test-runner/wdio.conf.js]: A little history here, I added the config directory because I was mad that everything in src was getting transpiled. I didn't realize you could setup babel config to ignore folders. I think i'd propose, as a compromise, to place the config folder inside the lib folder.

**Review by mjhenkes — APPROVED** [review]

**Review by emilyrohrbough — APPROVED** [review]

**Pull request merged**
